File: sdks/python/cloudpool/services/compute.py
"""Compute service (serverless, containers, cron)."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from cloudpool._client import CloudPoolClient
from cloudpool.models.compute import (
    ContainerDeployment,
    CronExecution,
    CronJob,
    ServerlessFunction,
    StaticSite,
)


class ComputeClient:
    """Synchronous compute client.

    Manages static sites, serverless functions, container deployments,
    and cron jobs.

    Accessed via ``cloudpool.compute`` on a ``CloudPool`` instance.
    """

    def __init__(self, client: CloudPoolClient) -> None:
        self._client = client

    # ── Static Sites ──

    def deploy_static_site(self, name: str, bucket_name: str, domain: str) -> StaticSite:
        """Deploy a static site.

        Args:
            name: Site name.
            bucket_name: Source bucket for site files.
            domain: Custom domain or subdomain.

        Returns:
            The deployed StaticSite.
        """
        resp = self._client.request(
            "POST", "/api/compute/static",
            json={"name": name, "bucketName": bucket_name, "domain": domain},
        )
        return StaticSite.from_dict(resp)

    def list_static_sites(self) -> List[StaticSite]:
        """List static sites.

        Returns:
            List of StaticSite objects.
        """
        resp = self._client.request("GET", "/api/compute/static")
        return [StaticSite.from_dict(s) for s in (resp if isinstance(resp, list) else [])]

    def delete_static_site(self, site_id: str) -> None:
        """Delete a static site.

        Args:
            site_id: The site's unique identifier.
        """
        self._client.request("DELETE", f"/api/compute/static/{site_id}")

    # ── Serverless Functions ──

    def deploy_serverless_function(
        self,
        name: str,
        trigger_route: str,
        code: str,
        runtime: str = "node18",
    ) -> ServerlessFunction:
        """Deploy a serverless function.

        Args:
            name: Function name.
            trigger_route: HTTP trigger path.
            code: Function source code.
            runtime: Runtime environment (e.g., "node18", "python3").

        Returns:
            The deployed ServerlessFunction.
        """
        resp = self._client.request(
            "POST", "/api/compute/serverless",
            json={
                "name": name,
                "triggerRoute": trigger_route,
                "code": code,
                "runtime": runtime,
            },
        )
        return ServerlessFunction.from_dict(resp)

    def list_serverless_functions(self) -> List[ServerlessFunction]:
        """List serverless functions.

        Returns:
            List of ServerlessFunction objects.
        """
        resp = self._client.request("GET", "/api/compute/serverless")
        return [ServerlessFunction.from_dict(f) for f in (resp if isinstance(resp, list) else [])]

    def execute_serverless_function(
        self,
        function_id: str,
        params: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Execute a serverless function.

        Args:
            function_id: The function's unique identifier.
            params: Optional input parameters.

        Returns:
            Function execution result.
        """
        return self._client.request(
            "POST", f"/api/compute/serverless/{function_id}/execute",
            json=params or {},
        )

    def delete_serverless_function(self, function_id: str) -> None:
        """Delete a serverless function.

        Args:
            function_id: The function's unique identifier.
        """
        self._client.request("DELETE", f"/api/compute/serverless/{function_id}")

    # ── Container Deployments ──

    def deploy_container(
        self,
        name: str,
        docker_image: str,
        cpu: str = "0.5",
        memory: str = "512Mi",
        replicas: int = 1,
    ) -> ContainerDeployment:
        """Deploy a container.

        Args:
            name: Deployment name.
            docker_image: Docker image URL.
            cpu: CPU allocation (e.g., "0.5", "1").
            memory: Memory allocation (e.g., "512Mi", "1Gi").
            replicas: Number of replicas.

        Returns:
            The deployed ContainerDeployment.
        """
        resp = self._client.request(
            "POST", "/api/compute/container",
            json={
                "name": name,
                "dockerImage": docker_image,
                "cpu": cpu,
                "memory": memory,
                "replicas": replicas,
            },
        )
        return ContainerDeployment.from_dict(resp)

    def list_containers(self) -> List[ContainerDeployment]:
        """List container deployments.

        Returns:
            List of ContainerDeployment objects.
        """
        resp = self._client.request("GET", "/api/compute/container")
        return [ContainerDeployment.from_dict(c) for c in (resp if isinstance(resp, list) else [])]

    def delete_container(self, container_id: str) -> None:
        """Delete a container deployment.

        Args:
            container_id: The container's unique identifier.
        """
        self._client.request("DELETE", f"/api/compute/container/{container_id}")

    # ── Cron Jobs ──

    def list_cron_jobs(self, project_id: str) -> List[CronJob]:
        """List cron jobs for a project.

        Args:
            project_id: The project's unique identifier.

        Returns:
            List of CronJob objects.
        """
        resp = self._client.request("GET", f"/api/v1/projects/{project_id}/cron")
        return [CronJob.from_dict(j) for j in (resp if isinstance(resp, list) else [])]

    def create_cron_job(
        self,
        project_id: str,
        name: str,
        cron_expression: str,
        target_url: str,
        http_method: str = "GET",
        payload: Optional[str] = None,
        headers: Optional[Dict[str, str]] = None,
        is_active: bool = True,
    ) -> CronJob:
        """Create a cron job.

        Args:
            project_id: The project's unique identifier.
            name: Job name.
            cron_expression: Cron schedule expression.
            target_url: HTTP endpoint URL.
            http_method: HTTP method (GET, POST, etc.).
            payload: Optional request body.
            headers: Optional request headers.
            is_active: Whether the job starts active.

        Returns:
            The created CronJob.
        """
        body: Dict[str, Any] = {
            "name": name,
            "cronExpression": cron_expression,
            "targetUrl": target_url,
            "httpMethod": http_method,
            "isActive": is_active,
        }
        if payload:
            body["payload"] = payload
        if headers:
            body["headers"] = headers
        resp = self._client.request(
            "POST", f"/api/v1/projects/{project_id}/cron", json=body,
        )
        return CronJob.from_dict(resp)

    def delete_cron_job(self, project_id: str, job_id: str) -> None:
        """Delete a cron job.

        Args:
            project_id: The project's unique identifier.
            job_id: The job's unique identifier.
        """
        self._client.request("DELETE", f"/api/v1/projects/{project_id}/cron/{job_id}")

    def get_cron_executions(self, project_id: str, job_id: str) -> List[CronExecution]:
        """Get execution history for a cron job.

        Args:
            project_id: The project's unique identifier.
            job_id: The job's unique identifier.

        Returns:
            List of CronExecution objects.
        """
        resp = self._client.request(
            "GET", f"/api/v1/projects/{project_id}/cron/{job_id}/executions",
        )
        return [CronExecution.from_dict(e) for e in (resp if isinstance(resp, list) else [])]


class AsyncComputeClient:
    """Asynchronous compute client."""

    def __init__(self, client: Any) -> None:
        self._client = client

    async def deploy_serverless_function(self, name: str, trigger_route: str, code: str) -> ServerlessFunction:
        resp = await self._client.request("POST", "/api/compute/serverless", json={"name": name, "triggerRoute": trigger_route, "code": code})
        return ServerlessFunction.from_dict(resp)

    async def list_serverless_functions(self) -> List[ServerlessFunction]:
        resp = await self._client.request("GET", "/api/compute/serverless")
        return [ServerlessFunction.from_dict(f) for f in (resp if isinstance(resp, list) else [])]

    async def execute_serverless_function(self, function_id: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return await self._client.request("POST", f"/api/compute/serverless/{function_id}/execute", json=params or {})

    async def delete_serverless_function(self, function_id: str) -> None:
        await self._client.request("DELETE", f"/api/compute/serverless/{function_id}")

    async def list_cron_jobs(self, project_id: str) -> List[CronJob]:
        resp = await self._client.request("GET", f"/api/v1/projects/{project_id}/cron")
        return [CronJob.from_dict(j) for j in (resp if isinstance(resp, list) else [])]

    async def create_cron_job(self, project_id: str, name: str, cron_expression: str, target_url: str, http_method: str = "GET") -> CronJob:
        resp = await self._client.request("POST", f"/api/v1/projects/{project_id}/cron", json={
            "name": name, "cronExpression": cron_expression, "targetUrl": target_url, "httpMethod": http_method,
        })
        return CronJob.from_dict(resp)

    async def delete_cron_job(self, project_id: str, job_id: str) -> None:
        await self._client.request("DELETE", f"/api/v1/projects/{project_id}/cron/{job_id}")

File: sdks/go/compute.go
package cloudpool

import (
	"context"
	"strconv"
)

type ComputeClient struct {
	client *CloudPoolClient
}

type ServerlessDeployment struct {
	ID     string `json:"id"`
	Name   string `json:"name"`
	Status string `json:"status"`
	URL    string `json:"url,omitempty"`
}

type CronJob struct {
	ID       string `json:"id"`
	Name     string `json:"name"`
	Schedule string `json:"schedule"`
}

type PodInfo struct {
	ID     string `json:"id"`
	Image  string `json:"image"`
	Status string `json:"status"`
	CPU    string `json:"cpu,omitempty"`
	Memory string `json:"memory,omitempty"`
}

type LogEntry struct {
	Timestamp string `json:"timestamp"`
	Level     string `json:"level"`
	Message   string `json:"message"`
}

func (c *ComputeClient) DeployServerless(ctx context.Context, name, runtime string) (*ServerlessDeployment, error) {
	var resp ServerlessDeployment
	err := c.client.PostJSON(ctx, "/api/compute/serverless", map[string]interface{}{
		"name": name, "runtime": runtime,
	}, nil, &resp)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

func (c *ComputeClient) UploadCode(ctx context.Context, name, filePath string) (map[string]interface{}, error) {
	data, err := c.client.Upload(ctx, "/api/compute/serverless/"+name+"/code", filePath, "serverless")
	if err != nil {
		return nil, err
	}
	var resp map[string]interface{}
	if err := unmarshalResponse(data, &resp); err != nil {
		return nil, err
	}
	return resp, nil
}

func (c *ComputeClient) CreateCron(ctx context.Context, name, schedule string, code string) (*CronJob, error) {
	body := map[string]interface{}{"name": name, "schedule": schedule}
	if code != "" {
		body["code"] = code
	}
	var resp CronJob
	err := c.client.PostJSON(ctx, "/api/compute/cron", body, nil, &resp)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

func (c *ComputeClient) Deploy(ctx context.Context, directory string) (map[string]interface{}, error) {
	var resp map[string]interface{}
	err := c.client.PostJSON(ctx, "/api/compute/deploy", map[string]string{
		"path": directory,
	}, nil, &resp)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

func (c *ComputeClient) Logs(ctx context.Context, functionID string, tail int) ([]LogEntry, error) {
	var resp []LogEntry
	err := c.client.GetJSON(ctx, "/api/compute/"+functionID+"/logs", map[string]string{
		"tail": strconv.Itoa(tail),
	}, &resp)
	if err != nil {
		return nil, err
	}
	return resp, nil
}

func (c *ComputeClient) ListPods(ctx context.Context) ([]PodInfo, error) {
	var resp []PodInfo
	err := c.client.GetJSON(ctx, "/api/compute/pods", nil, &resp)
	if err != nil {
		return nil, err
	}
	return resp, nil
}
